### DISCUSSION TITLE
Sharing caches from runs-on/cache across multiple GitHub repos

### DISCUSSION THREAD

**Discussion by ruffsl** [runs-on/runs-on#74] [Ideas]
By default, I believe the scope for `runs-on/cache` replicates the same restrictions as `actions/cache` for accessing a cache. This is a simple way to enforce control flow for the provenance and propagation of cached assets, e.g. for cache isolation to avoid unintended cross talk between PRs or for security to avoid cache poisoning from untrusted or public PRs.

- https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache

However, for private orgs using poly repo source layouts, it would be nice to selectively specify for CI workflows for one repo to restore caches from another repo. Take for example the case where we have one integration repo and many smaller feature repos on a private GitHub org. The CI for the integration repo regularly rebuilds everything, including all other feature repos. While the CI for feature repos usually only build themselves and some subset of other feature repos.

When occasionally opening a PR to a feature repo, it would be nice to have the cache action fallback and restore from a target branch from the remote integration repo, rather than starting from a cold cache if such were locally evicted due to last accessed age, or divergence from upstream. This could be made to mimic the existing levels of scope for branches, but on a multi-repo level, with the integration repo being the ultimate fallback, rather than just the default branch on the feature repo itself.

I briefly looked into if it was feasible to replicate such control flow out-of-band by using symlinks in the S3 bucket, but that doesn't seem viable with S3 alone:

- https://stackoverflow.com/questions/29459885/does-amazon-s3-support-symlinks

Perhaps a reverse proxy could be used redirect requests with a given marker identifier, e.g. `::org_name/repo_name::`?

```diff
      - name: Cache ccache
        id: cache_ccache
        uses: runs-on/cache@v4
        with:
          save-always: true
          path: |
            ${{ inputs.overlay_ws }}/.ccache
          key: "ccache-v1\
            -${{ github.ref }}\
            -${{ github.run_id }}\
            -${{ github.run_number }}\
            -${{ github.run_attempt }}"
          restore-keys: |
            ccache-v1-${{ github.ref }}
            ccache-v1-refs/heads/${{ github.head_ref }}
            ccache-v1-refs/heads/${{ github.base_ref }}
+           ::org_name/repo_name::ccache-v1-refs/heads/${{ github.head_ref }}
+           ::org_name/repo_name::ccache-v1-refs/heads/${{ github.base_ref }}
```

I'm not insisting this be officially supported, but it would be nice to implement on my own AWS backend.

**Comment by crohr** [comment]
So `runs-on/cache` doesn't replicate the permission restrictions found on the official GHA cache. But the [Magic Cache](https://runs-on.com/caching/magic-cache/) that came around more recently does. So I think both can be used to serve different purposes, and I don't think I will implement cache restrictions depending on scopes for `runs-on/cache`.

**Reply by danielsantiago** [reply]
I think this should be better documented. We just finished migrating from `runs-on/cache` to Magic Cache and realized we can't share cache between PRs
